README
Tomboy version 0.01
===================

Back up your notes before using this! This software is in development, yada yada yada. 


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Net::DBus::GLib
  Class::Accessor

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2011 by David Collins

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Tomboy.

0.01  Sun Sep 27 23:56:23 2009
	- original version; created by h2xs 1.23 with options
		-AXc -n Tomboy



lib/Tomboy/Note.pm
package Tomboy::Note;

use Net::DBus::GLib;
use base qw(Class::Accessor);

Tomboy::Note->mk_ro_accessors(qw(uri));

use 5.010000;
use strict;
use warnings;
our $VERSION = '0.01';

our $white_list = {
    # process args for these methods
    # all of these methods require that the notes uri
    # be passed first which is being handled internally
    '__processArgs' => {
        'HideNote'              => 1,
        'DisplayNote'           => 1,
        'NoteExists'            => 1,
        'GetNoteChangeDate'     => 1,
        'GetNoteCreateDate'     => 1,
        'GetNoteTitle'          => 1,
        'DeleteNote'            => 1,
        'GetNoteContents'       => 1,
        'GetNoteContentsXml'    => 1,
        'GetNoteCompleteXml'    => 1,
        'SetNoteContents'       => 1,
        'SetNoteContentsXml'    => 1,
        'SetNoteCompleteXml'    => 1,
        'GetTagsForNote'        => 1,
        'AddTagToNote'          => 1,
        'RemoveTagFromNote'     => 1,
        'NoteDeleted'           => 1,
        'NoteAdded'             => 1,
        'NoteSaved'             => 1,
    }
};

# for calls that feel a bit more comfortable
our $aliases = {
    'addTag' => 'AddTagToNote',
    'content' => 'GetNoteContents',
    'createdDate' => 'GetNoteCreateDate',
    'changedDate' => 'GetNoteChangeDate',
    'delete' => 'DeleteNote',
    'exists' => 'NoteExists',
    'getTags' => 'GetTagsForNote',
    'title'  => 'GetNoteTitle',
};

sub new {
    my ($class, $args) = @_;
    my $bus     = Net::DBus::GLib->session;
    my $service = $bus->get_service("org.gnome.Tomboy");
    my $obj     = $service->get_object("/org/gnome/Tomboy/RemoteControl");
    
    my $self = bless {}, $class;
    
    if ($$args{uri}) {
        return 0 if !$obj->NoteExists($$args{uri});
        $$self{uri}  = $$args{uri};
    } else {
        $self->_createNote($obj, $args);
    }
    
    $$self{_obj} = $obj;
    
    return $self;
}

sub _createNote {
    my ($self, $obj, $args) = @_;
    my $note_title = $$args{title};
    my $note_content = $$args{content};
    # TODO: create this through the Tomboy OBJ
    $$self{uri} = $obj->CreateNamedNote($note_title);
    die "Could not create note with title: $note_title\n" if !$$self{uri};
    
    # currently a Tomboy bug preventing SetNoteContents from working via DBus
    # without displaying the note
    # https://bugzilla.gnome.org/show_bug.cgi?id=553062    
    $obj->SetNoteContentsXml($$self{uri},qq|<note-content version="0.1">$note_title\n\n$note_content</note-content>|);
}

sub AUTOLOAD {
    my ($self,@args) = @_;
    our ($AUTOLOAD);
    my $method = $AUTOLOAD;
    $method =~ s/.*:://;

    # check for an alias
    if (my $unaliased = $$aliases{$method}) {
        $method = $unaliased;
    }
    
    $self->{_obj}->$method if $$white_list{$method};
    
    my $processedMethods = $$white_list{'__processArgs'};
    if ($$processedMethods{$method}) {
        # prepare the arguments
        my $notes = [];
        
        if (@args) {
            # passing an array like this looks ugly
            # but it allows the values within it
            # to each be passed as seperate arguments
            unshift @args, $self->{uri}; 
            return $self->{_obj}->$method(@args);
        } else {
            return $self->{_obj}->$method($self->{uri});
        }
    }
}

1;
__END__



lib/Tomboy.pm
package Tomboy;

use Net::DBus::GLib;
use 5.010000;
use strict;
use warnings;
use Tomboy::Note;
our $VERSION = '0.01';

our $blacklist = {
    'CreateNote'      => 1,
    'CreateNamedNote' => 1,
};

our $white_list = {
    'Version'               => 1,
    'Introspect'            => 1,
    'DisplayNoteWithSearch' => 1,
    'FindStartHereNote'     => 1,
    'DisplaySearch'         => 1,
    'DisplaySearchWithText' => 1,
    # process args for these methods
    # to allow returning of Tomboy::Note Objects
    '__processArgs' => {
        'ListAllNotes'       => 1,
        'GetAllNotesWithTag' => 1,
        'SearchNotes'        => 1,
        'FindNote'           => 1,
    },
    '__creation' => {
        'CreateNote'            => 1,
        'CreateNamedNote'       => 1,
    }
};

# for calls that feel a bit more comfortable
our $aliases = {
    'findNote'              => 'FindNote',
    'searchNotes'           => 'SearchNotes',
    'getAllNotesWithTag'    => 'GetAllNotesWithTag',
    'findStartHereNote'     => 'FindStartHereNote',
    'listAllNotes'          => 'ListAllNotes',
};


sub new {
    my ($class,$args) = @_;
    
    my $bus = Net::DBus::GLib->session;
    my $service = $bus->get_service("org.gnome.Tomboy");
    
    return bless {
        _obj      => $service->get_object("/org/gnome/Tomboy/RemoteControl"),
    }, $class;
}

sub AUTOLOAD {
    my ($self,@args) = @_;
    our ($AUTOLOAD);
    my $method = $AUTOLOAD;
    $method =~ s/.*:://;
    
    # check for an alias
    if (my $unaliased = $$aliases{$method}) {
        $method = $unaliased;
    }
    
    # TODO: is this going to work or am i going to have
    # to further differentiate between methods that return
    # nothing and methods that have return values?
    return $self->{_obj}->$method if $$white_list{$method};
    
    # Methods that return a list of Tomboy::Note's
    my $processedMethods = $$white_list{'__processArgs'};
    if ($$processedMethods{$method}) {
        # prepare the arguments
        my $result = [];
        
        if (@args) {
            # passing an array like this looks ugly
            # but it allows the values within it
            # to each be passed as seperate arguments
            $result = $self->{_obj}->$method(@args);
        } else {
            $result = $self->{_obj}->$method;
        }

        # need to process result a bit
        if (ref($result)) {
            return [ map { Tomboy::Note->new({ uri => $_ }) } @$result ];
        } elsif ($result =~ qr{^note://tomboy/[\d\w-]*}) {
            # a single uri
            return Tomboy::Note->new({ uri => $result });
        }
    }

    # Creation methods
    my $creationMethods = $$white_list{'__creation'};
    if ($$creationMethods{$method}) {
        # creation methods should return a Tomboy::Note object
        my $uri = scalar @args ? 
            $self->{_obj}->$method(shift @args)
            :
            $self->{_obj}->$method
        ;
        
        return Tomboy::Note->new({ uri => $uri });
    }
}

1;
__END__

=head1 NAME

Tomboy - Perl API to access Tomboy functionality through DBus

=head1 SYNOPSIS

my $tomboy = Tomboy->new;

my $total_notes = $tomboy->ListAllNotes;

$uri = $$total_notes[0]->uri;

ref($$total_notes[0]); # Tomboy::Note

my $notes = $tomboy->GetAllNotesWithTag($note_tag);

my $case_sensitive = 0;
my $search_notes = $tomboy->searchNotes("search term", $case_sensitive);

=head1 DESCRIPTION

This class is meant to go with Tomboy::Note. Note manipulation should be done there.
Tomboy MUST be installed to use this. This module uses DBus to access Tomboy's functionality. 
This module has NOT been tested on Windows or Mac. Software is still in development
so please back up notes before using just to be safe. 

=cut
